metrics-exporter/src/bin/io_engine/collector/pool_stat.rs

```diff
@@ -0,0 +1,185 @@
+    pool_read_latency_us: GaugeVec,
+    pool_write_latency_us: GaugeVec,
+    descs: Vec<Desc>,
+}
+
+impl Default for PoolIoStatsCollector {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+/// Initialize all the metrics to be defined for pools iostat collector.
+impl PoolIoStatsCollector {
+    /// Initialize all the metrics to be defined for pools iostat collector.
+    pub fn new() -> Self {
+        let mut descs = Vec::new();
+
+        let pool_bytes_read =
+            init_gauge_vec("bytes_read", "Total bytes read on the pool", &mut descs);
+        let pool_num_read_ops = init_gauge_vec(
+            "num_read_ops",
+            "Number of read operations on the pool",
+            &mut descs,
+        );
+        let pool_bytes_written = init_gauge_vec(
+            "bytes_written",
+            "Total bytes written on the pool",
+            &mut descs,
+        );
+        let pool_num_write_ops = init_gauge_vec(
+            "num_write_ops",
+            "Number of write operations on the pool",
+            &mut descs,
+        );
+        let pool_read_latency_us = init_gauge_vec(
+            "read_latency_us",
+            "Total read latency on the pool in usec",
+            &mut descs,
+        );
+        let pool_write_latency_us = init_gauge_vec(
+            "write_latency_us",
+            "Total write latency on the pool in usec",
+            &mut descs,
+        );
+
+        Self {
+            pool_bytes_read,
+            pool_num_read_ops,
+            pool_bytes_written,
+            pool_num_write_ops,
+            pool_read_latency_us,
+            pool_write_latency_us,
+            descs,
+        }
+    }
+}
+
+impl Collector for PoolIoStatsCollector {
+    fn desc(&self) -> Vec<&Desc> {
+        self.descs.iter().collect()
+    }
+
+    fn collect(&self) -> Vec<prometheus::proto::MetricFamily> {
+        let c = match Cache::get_cache().lock() {
+            Ok(c) => c,
+            Err(error) => {
+                error!(%error,"Error while getting cache resource");
+                return Vec::new();
+            }
+        };
+        let cp = c.deref();
+        let mut metric_family = Vec::with_capacity(6 * cp.pool_iostat().pool_stats.capacity());
+        let node_name = match get_node_name() {
+            Ok(name) => name,
+            Err(error) => {
+                error!(?error, "Unable to get node name");
+                return metric_family;
+            }
+        };
+
+        for i in &cp.pool_iostat().pool_stats {
+            let p: &PoolIoStat = i;
+
+            let pool_bytes_read = match self
+                .pool_bytes_read
+                .get_metric_with_label_values(&[node_name.clone().as_str(), p.name().as_str()])
+            {
+                Ok(pool_bytes_read) => pool_bytes_read,
+                Err(error) => {
+                    error!(%error, "Error while creating metrics(pool_bytes_read) with label values");
+                    return metric_family;
```

Comment:
I don't quite follow how this api works as this doesn't seem to return an error, but rather the vector.
Is there any way to return error? If not, should we just carry on and try to fill as much info as we can?